salt/state.py

```diff
@@ -3391,13 +3419,9 @@ def _filter_matches(_match, _data, _opts):
                                     _filter_matches(match, data, _opts)
                             if isinstance(item, six.string_types):
                                 matches[saltenv].append(item)
-                            elif isinstance(item, dict):
-                                env_key, inc_sls = item.popitem()
-                                if env_key not in self.avail:
-                                    continue
-                                if env_key not in matches:
-                                    matches[env_key] = []
-                                matches[env_key].append(inc_sls)
+                            # Pass through any keyword parameters.
+                            if isinstance(item, dict):
+                                matches[saltenv].append(item)
```

Comment:
I also don't understand this. The code you're replacing looks to be designed to include matches from a specific environment, and your replacement code removes this completely. Can you explain your thinking here?

Reply:
To me it *appeared* to do that as well, however, I'm not really sure if it works as expected (I don't claim to have knowledge or even justified belief in this case).

When the code goes through a list of matches produced by an instance of `salt.minion.Matcher` in a for loop (`for item in _data`), each loop variable should be either a string naming a state or a dictionary stating a key-value pair or pairs to specify keyword parameters (e.g. `{ 'match': 'pcre' }`).

If that is really the case, in what situation the loop variable would have a value suitable for popping `env_key, inc_sls`? That'd require a line like `someenv:somestate`, which seems like an entry not suitable for the top file.

I'm not claiming that I've exhaustively *tested* every possible scenario, and I may have misunderstood the format of the top file, in which case there might very well be other kind of values than just the two I mentioned above.

But, should I've grasped it correctly (and I doubt I am, that is why sent the PR to beg some comments), lines from `env_key, inc_sls = item.popitem()` to `matches[env_key].append(inc_sls)` don't do what it appears to be trying to accomplish, instead a single line `matches[saltenv].append(item)` would then do what is desired ie. take whatever is in the dictionary and pass it forward untouched.